kernel: mm: only include demand_paging.h if needed

dcpleung opened the pull request:
This moves including of demand_paging.h out of kernel/mm.h,
so that users of demand paging APIs must include the header
explicitly. Since the main user is kernel itself, we can be
more discipline about header inclusion.

dcpleung committed:
sys: arch_interface: include arch/syscall.h if USERSPACE

Due to arch_syscall_invoke are declared static, if there are
no corresponding definition, compiler will complain. It has
been working fine so far because arch/syscall.h is included
indirectly through some other headers (especially through
generated syscall stubs). However, with enough header files
shuffling this becomes an issue. So include arch/syscall.h
in arch_interface.h explicitly.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

andyross reviewed (commented):
Is there a reason for the header to be transitive in the first place?  Generally with an optional API, you make the users (who have to turn on the kconfig) include it explicitly, as by definition nothing depends on it.

(I guess one answer is that headers are APIs and we're stuck with it, but IIRC demand paging isn't a stable API yet, so we're safe?)

dcpleung committed:
kernel: mm: only include demand_paging.h if needed

This moves including of demand_paging.h out of kernel/mm.h,
so that users of demand paging APIs must include the header
explicitly. Since the main user is kernel itself, we can be
more discipline about header inclusion.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
> Is there a reason for the header to be transitive in the first place? Generally with an optional API, you make the users (who have to turn on the kconfig) include it explicitly, as by definition nothing depends on it.
> 
> (I guess one answer is that headers are APIs and we're stuck with it, but IIRC demand paging isn't a stable API yet, so we're safe?)

That actually is a better idea. I've made the changes to include `demand_paging.h` only if needed. Hopefully CI will pass.
